test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1

```diff
@@ -0,0 +1,917 @@
+
+        Import-Module -Name $moduleDirectory\TestCmdletForConstrainedLanguage.dll
+    }
+} # end if ($IsWindows)
+
+try
+{
+    $defaultParamValues = $PSDefaultParameterValues
+    $PSDefaultParameterValues["it:Skip"] = !$IsWindows
+
+    Describe "Built-ins work within constrained language" -Tags 'Feature','RequireAdminOnWindows' {
+
+        BeforeAll {
+            $TestCasesBuiltIn = @(
+                @{testName = "Verify built-in function"; scriptblock = { Get-Verb } }
+                @{testName = "Verify build-in error variable"; scriptblock = { Write-Error SomeError -ErrorVariable ErrorOutput -ErrorAction SilentlyContinue; $ErrorOutput } }
+            )
+        }
+
+        It "<testName>" -TestCases $TestCasesBuiltIn {
+
+            param ($scriptblock)
+
+            try
+            {
+                Invoke-LanguageModeTestingSupportCmdlet -EnableConstrainedLanguageMode
+
+                $result = (& $scriptblock)
+            }
+            finally
+            {
+                Invoke-LanguageModeTestingSupportCmdlet -EnableFullLanguageMode
+            }
+
+            ($result.Count -gt 0) | Should Be $true
+        }
+    }
+
+    Describe "Background jobs" -Tags 'Feature','RequireAdminOnWindows' {
+
+        Context "Background jobs in system lock down mode" {
+
+            # TODO
+            # Test is pending because PowerShell Core currently stubs out Windows system lockdown APIs
+            It "Verifies that background jobs in system lockdown mode run in constrained language" -Pending {
+
+                try
+                {
+                    Invoke-LanguageModeTestingSupportCmdlet -SetLockdownMode
+
+                    $job = Start-Job -ScriptBlock { [object]::Equals("A", "B") } | Wait-Job
+                    $expectedError = $job.ChildJobs[0].Error
+                    $job | Remove-Job
+                }
+                finally
+                {
+                    Invoke-LanguageModeTestingSupportCmdlet -RevertLockdownMode
+                }
+
+                $expectedError | Should Not Be $null
+                $expectedError.FullyQualifiedErrorId | Should Match "MethodInvocationNotSupportedInConstrainedLanguage"
+            }
+        }
+
+        Context "Background jobs within inconsistent mode" {
+
+            It "Verifies that background job is denied when mode is inconsistent" {
+
+                try
+                {
+                    Invoke-LanguageModeTestingSupportCmdlet -EnableConstrainedLanguageMode
+
+                    Start-Job { [object]::Equals("A", "B") }
+                    throw "No Exception!"
+                }
+                catch 
+                {
+                    $exception = $_
+                }
+                finally
+                {
+                    Invoke-LanguageModeTestingSupportCmdlet -EnableFullLanguageMode
+                }
+
+                $exception.FullyQualifiedErrorId | Should Match "CannotStartJobInconsistentLanguageMode"

```

Comment:
Since Start-Job throws, you should use ShouldBeErrorId

Reply:
No, direct match to ErrorId is fragile because the provided Id can be augmented over time.  I prefer a match test.

Reply:
Ok, can you add a comment then since you're using a different pattern than what we recommend?

Reply:
I don't think we should recommend this pattern unless we require FullyQualifiedErrorId to always be unchanging.  But this wasn't true in the past and as far as I know the only requirement is that it be unique and readable.  I have investigated numerous test failures where the ErrorId was augmented (sometimes by rolling up nested errors via method calls or the remoting system) and so the direct comparison failed even though the test succeeded and the error Id contained the correct error.  It might be better change ShouldBeErrorId to do a contains rather than an exact match, or create a new "ShouldContainErrorId" keyword.

Reply:
According to https://msdn.microsoft.com/en-us/library/ms714465(v=vs.85).aspx the FQErrorId should be reproducible and same across systems, so not sure what you mean by augmented.

Reply:
What I mean by augmented is that the FQErrorId string is added to under some circumstances as errors are rolled up.  For example you may throw an error with error string "CannotDefineSupportedCommand", but the actual error Id you get has more information, something like:
```powershell
"CannotDefineSupportedCommand,Microsoft.PowerShell.Commands.InvokeExpressionCommand"
or
"CannotDefineSupportedCommand,PSRemotingTransportException"
```
depending on how the error came about.

The problem is that in the past the reported FQErrorId string changed due to code refactoring or new features and our tests broke because they required exact matches on the entire FQErrorId string, even when the actual error component was still in the string:
```powershell
"CannotDefineSupportedCommand,MethodCallFailed,PSRemotingTransportException"
FQErrorID -eq "CannotDefineSupportedCommand,PSRemotingTransportException" -> Test Failure
```

That is why I advocate for tests to match the expected *component* in a FQErrorId string:
```powershell
FQErrorID -match "CannotDefineSupportedCommand"  -> Test Succeeds
```

Reply:
This seems to contradict the MSDN link above.  cc @BrucePay

Reply:
I don't think so.  PowerShell processing doesn't change or remove FQErrorIds for a specific code path, it just adds more information under certain circumstances.  You can still match to the specific error Id you care about.

Anyway I am not defending how it works, just relating my experience as to why I don't trust an exact match for tests.